sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -641,29 +622,20 @@ object ScalaJSPluginInternal {
       testHtml := {
         val log = streams.value.log
         val output = (artifactPath in testHtml).value
+        val title = name.value + " - tests"
+        val jsFiles = (jsExecutionFiles in testHtml).value
 
-        val jsFileCache = new VirtualFileMaterializer(true)
-        val jsFileURIs = (jsExecutionFiles in testHtml).value.map {
-          case file: FileVirtualFile => file.toURI
-          case file                  => jsFileCache.materialize(file).toURI
-        }
+        val frameworks = (loadedTestFrameworks in testHtml).value.toList
+        val frameworkImplClassNames =
+          frameworks.map(_._1.implClassNames.toList)
 
-        val css: java.io.File = {
-          val name = "test-runner.css"
-          val inputStream = getClass.getResourceAsStream(name)
-          try {
-            val outFile = (resourceManaged in testHtml).value / name
-            IO.transfer(inputStream, outFile)
-            outFile
-          } finally {
-            inputStream.close()
-          }
+        val taskDefs = for (td <- (definedTests in testHtml).value) yield {
+          new sbt.testing.TaskDef(td.name, td.fingerprint,
+              td.explicitlySpecified, td.selectors)
         }
 
-        IO.write(output, HTMLRunnerTemplate.render(output.toURI,
-            name.value + " - tests", jsFileURIs, css.toURI,
-            (loadedTestFrameworks in testHtml).value,
-            (definedTests in testHtml).value))
+        HTMLRunnerBuilder.writeToFile(output, title, jsFiles,
+            frameworkImplClassNames, taskDefs.toList)
```

Comment:
The only non-trivial conflict was here, because of the change `testHtmlKey` -> `testHtml` in the base branch. But it was just a matter of taking the new implementation from this commit, and re-replacing `testHtmlKey` by `testHtml`.